### ISSUE TITLE
Text manipulation commands: bug reports, features suggestions, and other discussion

### ISSUE THREAD

**Issue by alexboche** [dictation-toolbox/Caster#579]

**Comment by alexboche** [comment]
I don't think these working Emacs yet since the selection behavior is different there.

**Comment by kendonB** [comment]
In Word, starting here:
![image](https://user-images.githubusercontent.com/5327505/59569033-03a52100-90d8-11e9-9f80-c5ef7c09f020.png)
and saying "go sauce before equals" ends up here:
![image](https://user-images.githubusercontent.com/5327505/59569053-3a7b3700-90d8-11e9-94a5-8001331fd85c.png)

**Comment by kendonB** [comment]
To improve the speed, could we use Ctrl-Right/Ctrl-Left for cursor movements to speed things up? The rules for what types of characters get skipped or stopped at seem the same in most editors. We could also (in-principle) use the up/down arrows for fixed width editors without line wrapping by detecting end line characters, I think.

**Comment by alexboche** [comment]
Sometimes people will want to search for more than just one letter because there are a lot of occurrences of say "a"so e.g. one might want to say something like "go lease sierra mike". I tried implementing that [here](https://paste.gg/p/anonymous/c3bf9598efb342e3a6dbe25b09fe0cf5) by making the choice object the set of sequences up to length 2 of the characters (alphabet and punctuation). 
There are two problems with this, both of which would be solved if these commands were not CCR (though I think that making these CCR has a lot of advantages). First,  
I encountered the grammar complexity error even with sequences of length 2. If we made the commands non-CCR, I'm guessing we could get up to sequences of length 4 without hitting grammar complexity error. Second, there is ambiguity about whether "go lease sierra mike" means go to the nearest occurrence of "sm" or go to the nearest occurrence of "m" and then press s. Both interpretations have situations in which they would be desirable.

**Comment by kendonB** [comment]
What if we had another word before these so it knows to look for length 2 characters? Like "go sauce 2 double mike sierra"?

**Comment by alexboche** [comment]
> To improve the speed, could we use Ctrl-Right/Ctrl-Left for cursor movements to speed things up? The rules for what types of characters get skipped or stopped at seem the same in most editors. We could also (in-principle) use the up/down arrows for fixed width editors without line wrapping by detecting end line characters, I think.

Good idea. I don't know when I'll have time to implement that. Would anybody else have time soon?

**Comment by kendonB** [comment]
I found differences in Ctrl-Right/Left behavior across Notepad++, GitHub on Chrome, and TexStudio (all 3 different). Small sample but it seems like there will be a lot of edge cases to deal with

**Comment by alexboche** [comment]
In the meantime, people could try turning down the keypress wait time in settings.toml though that doesn't seem to be the bottleneck in some places (e.g. in gitter chat)

**Comment by kendonB** [comment]
I did some quick profiling and found that whenever there's a decent lag it's because of the navigation using the arrow keys.

**Comment by LexiconCode** [comment]
@alexboche mentioned he was going to be busy for the next few weeks. Therefore I might unpin this issue if no one has a preference.

**Comment by kendonB** [comment]
It would be neat to be able to integrate these with aliases. Imagine you've got a piece of code with a particular phrase/variable that you have an alias for. For example in the thing I'm currently writing, I have aliased `g_{C}` with "GC". It would be nice to be able to say "go lease before GC".

**Comment by alexboche** [comment]
> I found differences in Ctrl-Right/Left behavior across Notepad++, GitHub on Chrome, and TexStudio (all 3 different). Small sample but it seems like there will be a lot of edge cases to deal with

Perhaps this could be implemented just for when the target object is dictation rather than characters. If everything is just words, then it should be pretty simple--basically one just counts the number of spaces. (with possibly a few exceptions). characters (e.g. in a programming context) would be much more complicated. 

The functions have the ability to do different stuff for each application, so one could start by implementing the Ctrl-Right/Left for a particular application and then having the functions do the regular arrow key method for all other applications until those are implemented.

**Comment by alexboche** [comment]
I was just using these a bit and the move/go commands were mistakenly interpreting some of the variable names as dictation such as before/after and 2nd,3rd as in `occurrence_number` .  When this happens, and it only happens intermittently, it causes the commands to not work. I don't understand what causes the problem to occur versus not. I can think of three things we could try to solve this. 

1) introduce new non-English words to play the role of "before" and "after" such as "bef" "aft". This might make the engine more likely to go down the character_sequence path rather than the dictation path. (I don't know if this will work. as I try to test it now, the problem is not occurring in the first place even without this.)

2) put in optional or required word such as "bow" in the __middle__ of the spec for the character sequence version of the command and not in the dictation version.

3) We can give different names (i.e. the __first__ word of spec) to the dictation in the character sequence versions of the commands. For example with go/move, we could have the character sequence command be pronounced either go or move but require the dictation version to be only pronounced move. That way users could use the same word for both if they are not running into problems but use different words if they are. If the problem proves to be consistent, we can make different words required.

I think the problem is mostly with the go/move commands since those tend to use the most variables, but this problem could arise for all of the commands. 
I've hardly gotten around to using these commands much since I just got around to installing them on my own personal branch today. Is anyone else using these commands? Have people experience this problem? @kendonB

**Comment by kendonB** [comment]
Can you spell out the exact utterances both that you're attempting and what is getting interpreted? I've had no trouble with misinterpretations like what you describe.

**Comment by alexboche** [comment]
I don't remember what was causing the problem for me previously, but another user called parashoot was reporting similar problems where character names were being interpreted as dictation (e.g. "arch" was being interpreted as the word arch instead of the letter a). Here are the [suggestions](https://gist.github.com/alexboche/fb3edf3a823744fb041733101331018c) that I gave him, and have not heard back other than that training punctuation names didn't work. I hope to look into this more fully at some point, but I don't have time now and I just want to put the linked suggestions here in a permanent location.

**Comment by Parashoot** [comment]
Hey That’s me, I found that I’ve gotten most of those issues resolved, by no fault of my own. So not sure how it fixed, the suggestions you gave did not do much to help me. But it did correct some of them on its own. @alexboche

**Comment by alexboche** [comment]
@Parashoot Thanks for reporting back, sorry my suggestions didn't help but glad it's working better now. Just to keep a record of things,  it would be useful if you could please confirm whether you tried the approaches where you actually changed the underlying code instead of just changing the names of characters or training command names. I previously called the approaches where you change the code 3 and 4, but i am now calling them 1 and 2 and have updated the linked [document](https://gist.github.com/alexboche/fb3edf3a823744fb041733101331018c)  accordingly. If you didn't try these that's no problem. I'm pretty sure these should work so if they didn't, there must've been a typo in my code.

**Comment by Parashoot** [comment]
@alexboche I did try change to the escape method but I found that it didn't improve the performance all that much since the program wasn't recognizing the characters properly. I have since had to rebuild my whole solution due to company firewall issues and updates clashing so it has reverted back to where it was before. LIke i mentioned it is able to pick up most things now except for comma, it almost seems as if the  ',' character it is looking for is different than the one inserted by typing a comma and it isn't matching because of a different utf character or whatever, not too sure, it could be thinking its actually something like ascii1284 when it should have been ascii1285 for example - not actual values. Regarding using the excape and it matching words within words, this is actually helpful for me when it does it by accident since I have to use a lot of navigation to camelcase variable names in code and don't otherwise use the go to "word" commands

**Comment by Parashoot** [comment]
Also after taking the suggesiton of anot her user I changed the 'comma' command to ',' in the punctuation and token lists, this appears to have fixed all my problems and sped up my program quite a bit, especially for things like jump in, out, and back, I think it wasn't recognizing the commas and thus was taking longer that it should have to process the characters and execute commands, seems weird to me.

**Comment by LexiconCode** [comment]
An overhaul of many of the text manipulation commands is in the pipeline mostly in the backend. The clipboard method will be improved upon and it also work with Dragonfly's accessibility API transparently when available.

**Comment by tieTYT** [comment]
I don't know if this is something that castor can fix, but after you use the capital command, the next thing you say will not have a space before it. If I didn't use capital command, the next thing I say would have a space before it.

In other words, if I say, "a b c capital up a d" it will print out like this: "A b cd"  I think it should print out like this: "A b c d"

**Comment by tieTYT** [comment]
can I request that the capital command toggles case by default? In other words if I tell it the search for a capitalized word, it will lower case that first letter

If I say "She said hi capital up she" I want the output to be "she said hi"

**Comment by tieTYT** [comment]
Imagine you have a sentence like this, "the fox went to the farm."

What I like about grab up, it looks for the previous word each time you say it. If your cursor is at the end of the sentence in you say "grab up the", the first time you say it, it will select the second "the", in the second time you say it, it will grab the first "the". I like this behavior.

But if your cursor was at the beginning of the sentence and you say `grab down the`, it will find the same word over and over again. I wish it found the next one each time like the grab up does. I would like to request that as a feature

**Comment by alexboche** [comment]
## Unrelated to previous comments:
If anyone is working interested in working on this more, I would suggest directing efforts toward the accessibility api commands. Either by adding new commands to it e.g. for navigating to individual character (sequences) or by extended the range of apps to which it applies. I think the clipboard method is probably close to the limit of what can be achieved with that (though making the cursor move faster in the clipboard would be a big improvement-- slow cursor movement is one of the resaons accessibility is so much better). In the long run, a deep integration method like the accessiblity api is definitely better. (I'll post this to the issue for future reference)
